Extension/lib/content-script/websocket.js

```diff
@@ -0,0 +1,296 @@
+/**
+ * This file is part of Adguard Browser Extension (https://github.com/AdguardTeam/AdguardBrowserExtension).
+ *
+ * Adguard Browser Extension is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Adguard Browser Extension is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Adguard Browser Extension.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+var overrideWebSocket = function () {
+
+    'use strict';
+    var Wrapped = window.WebSocket;
+    var toWrapped = new WeakMap();
+
+    var onResponseReceived = function (wrapper, ok) {
+
+        var bag = toWrapped.get(wrapper);
+        if (!ok) {
+            if (bag.properties.onerror) {
+                bag.properties.onerror(new window.ErrorEvent('error'));
+            }
+
+            return;
+        }
+
+        var wrapped = null;
+        try {
+            wrapped = new Wrapped(bag.args.url, bag.args.protocols);
+        } catch (ex) {
+            //Ignore ex
+        }
+
+        if (wrapped === null) {
+            return;
+        }
+
+        for (var p in bag.properties) {
+            wrapped[p] = bag.properties[p];
+        }
+        for (var i = 0, l; i < bag.listeners.length; i++) {
+            l = bag.listeners[i];
+            wrapped.addEventListener(l.ev, l.cb, l.fl);
+        }
+
+        toWrapped.set(wrapper, wrapped);
+    };
+
+    var noopfn = function () {};
+
+    var fallthruGet = function (wrapper, prop, value) {
+        var wrapped = toWrapped.get(wrapper);
+        if (!wrapped) {
+            return value;
+        }
+        if (wrapped instanceof Wrapped) {
+            return wrapped[prop];
+        }
+        return wrapped.properties.hasOwnProperty(prop) ?
+            wrapped.properties[prop] :
+            value;
+    };
+
+    var fallthruSet = function (wrapper, prop, value) {
+        if (value instanceof Function) {
+            value = value.bind(wrapper);
+        }
+        var wrapped = toWrapped.get(wrapper);
+        if (!wrapped) {
+            return;
+        }
+        if (wrapped instanceof Wrapped) {
+            wrapped[prop] = value;
+        } else {
+            wrapped.properties[prop] = value;
+        }
+    };
+
+    var WebSocket = function (url, protocols) {
+        'native';
+        if (window.location.protocol === 'https:'
+            && url.lastIndexOf('ws:', 0) === 0) {
+            var ws = new Wrapped(url, protocols);
+            if (ws) {
+                ws.close();
```

Comment:
I just don't understand, why is that so?
Forbidding WS from HTTPS pages or what?